### PR TITLE
Rewrite search requests on the coordinating nodes

### PR DESCRIPTION
**Pull request by s1monw**
This change rewrites search requests on the coordinating node before
we send requests to the individual shards. This will reduce the rewrite load
and object creation for each rewrite on the executing nodes and will fetch
resources only once instead of N times once per shard for queries like `terms`
query with index lookups. (among percolator and geo-shape)

Relates to #25791

### CODE REVIEW

**Comment by s1monw** [comment]
@clintongormley should we mention this in the migration docs?

**Comment by clintongormley** [comment]
@s1monw it's not breaking anything, so i don't think it needs a mention in the migration docs.

**Review by jimczi — APPROVED** [review]
LGTM

File: core/src/main/java/org/elasticsearch/search/SearchService.java [hunk 1]
```diff
@@ -935,6 +936,10 @@ public static boolean canRewriteToMatchNone(SearchSourceBuilder source) {
      * The action listener is guaranteed to be executed on the search thread-pool
      */
     private void rewriteShardRequest(ShardSearchRequest request, ActionListener<ShardSearchRequest> listener) {
+        // we also do rewrite on the coordinating node (TransportSearchService) but we also need to do it here for BWC as well as
+        // AliasFilters
+        // that might need to be rewritten. These are edge-cases but we are every efficient doing the rewrite here so it's not
+        // adding a lot of overhead
```
Comment (colings86) [inline core/src/main/java/org/elasticsearch/search/SearchService.java]: nit: strange formatting of comment?

**Review by colings86 — APPROVED** [review]
LGTM, I left one nit about the line wrapping of a comment.

File: core/src/test/java/org/elasticsearch/search/SearchServiceTests.java [hunk 2]
```diff
@@ -282,7 +282,10 @@ public FailOnRewriteQueryBuilder() {
 
         @Override
         protected QueryBuilder doRewrite(QueryRewriteContext queryShardContext) {
-            throw new IllegalStateException("Fail on rewrite phase");
+            if (queryShardContext.convertToShardContext() != null) {
```
Comment (jpountz) [inline core/src/test/java/org/elasticsearch/search/SearchServiceTests.java]: maybe rename queryShardContext to queryRewriteContext then to make it less confusing

Comment (s1monw) [inline core/src/test/java/org/elasticsearch/search/SearchServiceTests.java]: I pushed https://github.com/elastic/elasticsearch/commit/682abb90eed6ed3752bbc931f5a9e2aa6416e2bc to master for this

**Pull request merged**